File: compiler/tests/run_tiered_cpp98.py
#!/usr/bin/env python3
import argparse
import json
import os
import re
import subprocess
import sys
import tempfile
from datetime import datetime, timezone
from pathlib import Path


def parse_conformance_matrix(matrix_path: Path):
    entries = []
    line_rx = re.compile(r"^-\s+(.+):\s+(done|partial|missing)\s*$")
    for raw in matrix_path.read_text(encoding="utf-8").splitlines():
        m = line_rx.match(raw.strip())
        if not m:
            continue
        entries.append({"name": m.group(1), "status": m.group(2)})

    done = sum(1 for e in entries if e["status"] == "done")
    partial = sum(1 for e in entries if e["status"] == "partial")
    missing = sum(1 for e in entries if e["status"] == "missing")
    total = len(entries)
    weighted = done + 0.5 * partial
    weighted_pct = (100.0 * weighted / total) if total else 0.0

    return {
        "total": total,
        "done": done,
        "partial": partial,
        "missing": missing,
        "weighted": weighted,
        "weightedPct": round(weighted_pct, 1),
        "entries": entries,
    }


def run_case(case, repo_root: Path, tmp_dir: Path):
    cmd = [arg.replace("{tmpDir}", str(tmp_dir)) for arg in case["command"]]
    proc = subprocess.run(
        cmd,
        cwd=str(repo_root),
        capture_output=True,
        text=True,
    )

    errors = []

    expected_code = int(case.get("expectExitCode", 0))
    if proc.returncode != expected_code:
        errors.append(f"expected exit code {expected_code}, got {proc.returncode}")

    for snippet in case.get("stdoutContains", []):
        if snippet not in proc.stdout:
            errors.append(f"missing stdout snippet: {snippet}")

    for snippet in case.get("stderrContains", []):
        if snippet not in proc.stderr:
            errors.append(f"missing stderr snippet: {snippet}")

    return {
        "id": case["id"],
        "family": case.get("family", "unclassified"),
        "matrixFamily": case.get("matrixFamily"),
        "command": cmd,
        "exitCode": proc.returncode,
        "ok": len(errors) == 0,
        "errors": errors,
        "stdoutPreview": "\n".join(proc.stdout.splitlines()[-8:]),
        "stderrPreview": "\n".join(proc.stderr.splitlines()[-8:]),
    }


def run_tiers(plan, repo_root: Path):
    results = {"tier1": [], "tier2": [], "tier3": []}
    with tempfile.TemporaryDirectory(prefix="maiacpp_tiers_") as temp_dir:
        tmp_dir = Path(temp_dir)
        for tier_name in ["tier1", "tier2", "tier3"]:
            for case in plan["tiers"].get(tier_name, []):
                results[tier_name].append(run_case(case, repo_root, tmp_dir))
    return results


def summarize_tier(items):
    total = len(items)
    ok = sum(1 for i in items if i["ok"])
    failed = total - ok
    return {
        "total": total,
        "ok": ok,
        "failed": failed,
        "passRatePct": round((100.0 * ok / total), 1) if total else 0.0,
    }


def summarize_matrix_tracking(matrix, tier_results):
    matrix_names = [e["name"] for e in matrix.get("entries", [])]
    matrix_name_set = set(matrix_names)

    tracked = {}
    for tier_name in ["tier1", "tier2", "tier3"]:
        for item in tier_results.get(tier_name, []):
            fam = item.get("matrixFamily")
            if not fam or fam not in matrix_name_set:
                continue
            bucket = tracked.get(fam)
            if not bucket:
                bucket = {"family": fam, "cases": 0, "ok": 0, "failed": 0}
                tracked[fam] = bucket
            bucket["cases"] += 1
            if item.get("ok"):
                bucket["ok"] += 1
            else:
                bucket["failed"] += 1

    tracked_names = sorted(tracked.keys())
    untracked_names = [name for name in matrix_names if name not in tracked]

    return {
        "trackedFamilies": len(tracked_names),
        "totalFamilies": len(matrix_names),
        "trackedPct": round((100.0 * len(tracked_names) / len(matrix_names)), 1) if matrix_names else 0.0,
        "tracked": [tracked[name] for name in tracked_names],
        "untracked": untracked_names,
    }


def main():
    parser = argparse.ArgumentParser(description="Run MaiaCpp tiered C++98 suite and emit machine-readable report")
    parser.add_argument("--plan", default="compiler/tests/ebnf_tiers.json")
    parser.add_argument("--matrix", default="docs/CONFORMANCE_MATRIX.md")
    parser.add_argument("--out", default="out/reports/ebnf-tiered-report.json")
    args = parser.parse_args()

    repo_root = Path.cwd()
    plan_path = (repo_root / args.plan).resolve()
    matrix_path = (repo_root / args.matrix).resolve()
    out_path = (repo_root / args.out).resolve()

    if not plan_path.exists():
        print(f"[fail] plan not found: {plan_path}")
        return 2
    if not matrix_path.exists():
        print(f"[fail] matrix not found: {matrix_path}")
        return 2

    plan = json.loads(plan_path.read_text(encoding="utf-8"))
    matrix = parse_conformance_matrix(matrix_path)
    tier_results = run_tiers(plan, repo_root)

    summary = {
        "tier1": summarize_tier(tier_results["tier1"]),
        "tier2": summarize_tier(tier_results["tier2"]),
        "tier3": summarize_tier(tier_results["tier3"]),
    }
    matrix_tracking = summarize_matrix_tracking(matrix, tier_results)

    report = {
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "planPath": str(plan_path.relative_to(repo_root)),
        "matrixPath": str(matrix_path.relative_to(repo_root)),
        "matrixCoverage": {
            "totalFamilies": matrix["total"],
            "done": matrix["done"],
            "partial": matrix["partial"],
            "missing": matrix["missing"],
            "weightedImplementedPct": matrix["weightedPct"],
        },
        "tierSummary": summary,
        "matrixTracking": matrix_tracking,
        "tierResults": tier_results,
    }

    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(json.dumps(report, indent=2), encoding="utf-8")

    print("Tiered C++98 Report")
    print(f"- Matrix weighted coverage: {matrix['weightedPct']}% ({matrix['done']} done, {matrix['partial']} partial, {matrix['missing']} missing)")
    print(f"- Tier 1: {summary['tier1']['ok']}/{summary['tier1']['total']} passed")
    print(f"- Tier 2: {summary['tier2']['ok']}/{summary['tier2']['total']} passed")
    print(f"- Tier 3: {summary['tier3']['ok']}/{summary['tier3']['total']} passed")
    print(f"- Matrix families tracked by tier cases: {matrix_tracking['trackedFamilies']}/{matrix_tracking['totalFamilies']} ({matrix_tracking['trackedPct']}%)")
    print(f"- Report JSON: {out_path.relative_to(repo_root)}")

    # Non-zero when Tier 1 or Tier 2 fail. Tier 3 can fail while being tracked.
    if summary["tier1"]["failed"] > 0 or summary["tier2"]["failed"] > 0:
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: compiler/tests/run_tiered_cpp98.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const os = require('os');
const { spawnSync } = require('child_process');

function parseArgs(argv) {
  const out = {
    plan: 'compiler/tests/ebnf_tiers.json',
    matrix: 'docs/CONFORMANCE_MATRIX.md',
    out: 'out/reports/ebnf-tiered-report.json'
  };

  for (let i = 0; i < argv.length; i += 1) {
    const a = argv[i];
    if (a === '--plan' && i + 1 < argv.length) out.plan = argv[++i];
    else if (a === '--matrix' && i + 1 < argv.length) out.matrix = argv[++i];
    else if (a === '--out' && i + 1 < argv.length) out.out = argv[++i];
  }
  return out;
}

function parseConformanceMatrix(matrixPath) {
  const entries = [];
  const lineRx = /^-\s+(.+):\s+(done|partial|missing)\s*$/;
  const lines = fs.readFileSync(matrixPath, 'utf8').split(/\r?\n/);
  for (const raw of lines) {
    const m = lineRx.exec(raw.trim());
    if (!m) continue;
    entries.push({ name: m[1], status: m[2] });
  }

  const done = entries.filter((e) => e.status === 'done').length;
  const partial = entries.filter((e) => e.status === 'partial').length;
  const missing = entries.filter((e) => e.status === 'missing').length;
  const total = entries.length;
  const weighted = done + (0.5 * partial);
  const weightedPct = total ? Math.round((1000 * weighted / total)) / 10 : 0.0;

  return { total, done, partial, missing, weighted, weightedPct, entries };
}

function formatDurationMs(durationMs) {
  if (!Number.isFinite(durationMs) || durationMs < 0) return '0ms';
  if (durationMs < 1000) return `${durationMs}ms`;
  const seconds = durationMs / 1000;
  if (seconds < 60) return `${seconds.toFixed(seconds >= 10 ? 1 : 2)}s`;
  const minutes = Math.floor(seconds / 60);
  const remSeconds = seconds - (minutes * 60);
  return `${minutes}m${String(remSeconds.toFixed(1)).replace(/\.0$/, '')}s`;
}

function buildReport(planPath, matrixPath, matrix, tierResults) {
  const summary = {
    tier1: summarizeTier(tierResults.tier1),
    tier2: summarizeTier(tierResults.tier2),
    tier3: summarizeTier(tierResults.tier3)
  };
  const matrixTracking = summarizeMatrixTracking(matrix, tierResults);

  return {
    generatedAt: new Date().toISOString(),
    planPath,
    matrixPath,
    matrixCoverage: {
      totalFamilies: matrix.total,
      done: matrix.done,
      partial: matrix.partial,
      missing: matrix.missing,
      weightedImplementedPct: matrix.weightedPct
    },
    tierSummary: summary,
    matrixTracking,
    tierResults
  };
}

function writeReport(outPath, report) {
  fs.mkdirSync(path.dirname(outPath), { recursive: true });
  fs.writeFileSync(outPath, `${JSON.stringify(report, null, 2)}\n`, 'utf8');
}

function collectOutputFlagPaths(cmd) {
  const outputFlags = new Set([
    '--ast-json-out',
    '--ast-xml-out'
  ]);
  const outputs = [];
  for (let i = 0; i < cmd.length - 1; i += 1) {
    if (!outputFlags.has(cmd[i])) continue;
    outputs.push({ flag: cmd[i], path: cmd[i + 1] });
    i += 1;
  }
  return outputs;
}

function buildCacheKey(cmd) {
  if (cmd[0] !== 'node' || cmd[1] !== './compiler/cpp-compiler.js') {
    return null;
  }
  const normalized = [];
  for (let i = 0; i < cmd.length; i += 1) {
    normalized.push(cmd[i]);
    if ((cmd[i] === '--ast-json-out' || cmd[i] === '--ast-xml-out') && i + 1 < cmd.length) {
      normalized.push('<OUT>');
      i += 1;
    }
  }
  if (!normalized.includes('<OUT>')) {
    return null;
  }
  return normalized.join('\u0000');
}

function writeCachedArtifacts(cachedArtifacts, currentOutputs) {
  if (!Array.isArray(cachedArtifacts) || cachedArtifacts.length === 0) return;
  const byFlag = new Map(cachedArtifacts.map((artifact) => [artifact.flag, artifact]));
  for (const output of currentOutputs) {
    const cachedArtifact = byFlag.get(output.flag);
    if (!cachedArtifact) continue;
    fs.mkdirSync(path.dirname(output.path), { recursive: true });
    fs.writeFileSync(output.path, cachedArtifact.contents, 'utf8');
  }
}

function rewriteOutputPaths(text, cachedArtifacts, currentOutputs) {
  let out = String(text || '');
  if (!Array.isArray(cachedArtifacts) || !Array.isArray(currentOutputs)) {
    return out;
  }
  const currentByFlag = new Map(currentOutputs.map((entry) => [entry.flag, entry.path]));
  for (const artifact of cachedArtifacts) {
    const currentPath = currentByFlag.get(artifact.flag);
    if (!currentPath || artifact.path === currentPath) continue;
    out = out.split(artifact.path).join(currentPath);
  }
  return out;
}

function cloneCachedResult(cachedEntry, currentOutputs) {
  writeCachedArtifacts(cachedEntry.artifacts, currentOutputs);
  const stdout = rewriteOutputPaths(cachedEntry.stdout, cachedEntry.artifacts, currentOutputs);
  const stderr = rewriteOutputPaths(cachedEntry.stderr, cachedEntry.artifacts, currentOutputs);
  const stdoutLines = stdout.split(/\r?\n/);
  const stderrLines = stderr.split(/\r?\n/);

  return {
    exitCode: cachedEntry.exitCode,
    durationMs: 0,
    cacheHit: true,
    cachedDurationMs: cachedEntry.durationMs,
    ok: cachedEntry.ok,
    errors: [...cachedEntry.errors],
    stdout,
    stderr,
    stdoutPreview: stdoutLines.slice(Math.max(0, stdoutLines.length - 8)).join('\n'),
    stderrPreview: stderrLines.slice(Math.max(0, stderrLines.length - 8)).join('\n')
  };
}

function runCase(testCase, repoRoot, tmpDir, cache = new Map()) {
  const cmd = (testCase.command || []).map((arg) => String(arg).replace('{tmpDir}', tmpDir));
  const cacheKey = buildCacheKey(cmd);
  const currentOutputs = collectOutputFlagPaths(cmd);
  if (cacheKey && cache.has(cacheKey)) {
    const cached = cloneCachedResult(cache.get(cacheKey), currentOutputs);
    return {
      id: testCase.id,
      family: testCase.family || 'unclassified',
      matrixFamily: testCase.matrixFamily,
      command: cmd,
      ...cached
    };
  }

  const startedAt = Date.now();
  const proc = spawnSync(cmd[0], cmd.slice(1), { cwd: repoRoot, encoding: 'utf8' });
  const stdout = proc.stdout || '';
  const stderr = proc.stderr || '';
  const errors = [];

  const expectedCode = Number(testCase.expectExitCode || 0);
  if ((proc.status ?? 1) !== expectedCode) {
    errors.push(`expected exit code ${expectedCode}, got ${proc.status}`);
  }

  for (const snippet of (testCase.stdoutContains || [])) {
    if (!stdout.includes(snippet)) errors.push(`missing stdout snippet: ${snippet}`);
  }
  for (const snippet of (testCase.stderrContains || [])) {
    if (!stderr.includes(snippet)) errors.push(`missing stderr snippet: ${snippet}`);
  }

  const stdoutLines = stdout.split(/\r?\n/);
  const stderrLines = stderr.split(/\r?\n/);
  const artifacts = currentOutputs
    .filter((output) => fs.existsSync(output.path))
    .map((output) => ({
      flag: output.flag,
      path: output.path,
      contents: fs.readFileSync(output.path, 'utf8')
    }));

  const result = {
    id: testCase.id,
    family: testCase.family || 'unclassified',
    matrixFamily: testCase.matrixFamily,
    command: cmd,
    exitCode: proc.status,
    durationMs: Date.now() - startedAt,
    cacheHit: false,
    cachedDurationMs: null,
    ok: errors.length === 0,
    errors,
    stdout,
    stderr,
    stdoutPreview: stdoutLines.slice(Math.max(0, stdoutLines.length - 8)).join('\n'),
    stderrPreview: stderrLines.slice(Math.max(0, stderrLines.length - 8)).join('\n')
  };

  if (cacheKey && result.ok) {
    cache.set(cacheKey, {
      exitCode: result.exitCode,
      durationMs: result.durationMs,
      ok: result.ok,
      errors: [...result.errors],
      stdout,
      stderr,
      artifacts
    });
  }

  return result;
}

function runTiers(plan, repoRoot, options = {}) {
  const results = { tier1: [], tier2: [], tier3: [] };
  const reportContext = options.reportContext || null;
  const cache = new Map();
  const totalCasesByTier = {
    tier1: ((plan.tiers || {}).tier1 || []).length,
    tier2: ((plan.tiers || {}).tier2 || []).length,
    tier3: ((plan.tiers || {}).tier3 || []).length
  };
  const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'maiacpp_tiers_'));
  try {
    for (const tierName of ['tier1', 'tier2', 'tier3']) {
      const cases = ((plan.tiers || {})[tierName] || []);
      if (cases.length === 0) {
        console.log(`[tiered] ${tierName}: no cases`);
        continue;
      }
      console.log(`[tiered] ${tierName}: starting ${cases.length} case(s)`);
      for (let index = 0; index < cases.length; index += 1) {
        const testCase = cases[index];
        const label = `${tierName} ${index + 1}/${cases.length}`;
        console.log(`[tiered] ${label}: ${testCase.id}`);
        const result = runCase(testCase, repoRoot, tempDir, cache);
        results[tierName].push(result);
        const status = result.ok ? 'ok' : 'fail';
        const durationText = result.cacheHit
          ? `cached from ${formatDurationMs(result.cachedDurationMs)} baseline`
          : formatDurationMs(result.durationMs);
        console.log(`[tiered] ${label}: ${status} (${durationText})`);
        if (!result.ok) {
          for (const error of result.errors) {
            console.log(`[tiered] ${label}: ${error}`);
          }
        }
        if (reportContext) {
          const report = buildReport(
            reportContext.planPath,
            reportContext.matrixPath,
            reportContext.matrix,
            results
          );
          writeReport(reportContext.outPath, report);
        }
      }
      const tierSummary = summarizeTier(results[tierName]);
      console.log(
        `[tiered] ${tierName}: ${tierSummary.ok}/${totalCasesByTier[tierName]} passed (${tierSummary.passRatePct}%)`
      );
    }
  } finally {
    fs.rmSync(tempDir, { recursive: true, force: true });
  }
  return results;
}

function summarizeTier(items) {
  const total = items.length;
  const ok = items.filter((i) => i.ok).length;
  const failed = total - ok;
  const passRatePct = total ? Math.round((1000 * ok / total)) / 10 : 0.0;
  return { total, ok, failed, passRatePct };
}

function summarizeMatrixTracking(matrix, tierResults) {
  const matrixNames = (matrix.entries || []).map((e) => e.name);
  const matrixSet = new Set(matrixNames);
  const tracked = new Map();

  for (const tierName of ['tier1', 'tier2', 'tier3']) {
    for (const item of (tierResults[tierName] || [])) {
      const fam = item.matrixFamily;
      if (!fam || !matrixSet.has(fam)) continue;
      const bucket = tracked.get(fam) || { family: fam, cases: 0, ok: 0, failed: 0 };
      bucket.cases += 1;
      if (item.ok) bucket.ok += 1;
      else bucket.failed += 1;
      tracked.set(fam, bucket);
    }
  }

  const trackedNames = Array.from(tracked.keys()).sort();
  const untracked = matrixNames.filter((name) => !tracked.has(name));

  return {
    trackedFamilies: trackedNames.length,
    totalFamilies: matrixNames.length,
    trackedPct: matrixNames.length ? Math.round((1000 * trackedNames.length / matrixNames.length)) / 10 : 0.0,
    tracked: trackedNames.map((n) => tracked.get(n)),
    untracked
  };
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const repoRoot = process.cwd();
  const planPath = path.resolve(repoRoot, args.plan);
  const matrixPath = path.resolve(repoRoot, args.matrix);
  const outPath = path.resolve(repoRoot, args.out);

  if (!fs.existsSync(planPath)) {
    console.log(`[fail] plan not found: ${planPath}`);
    process.exit(2);
  }
  if (!fs.existsSync(matrixPath)) {
    console.log(`[fail] matrix not found: ${matrixPath}`);
    process.exit(2);
  }

  const plan = JSON.parse(fs.readFileSync(planPath, 'utf8'));
  const matrix = parseConformanceMatrix(matrixPath);
  const reportContext = {
    planPath: path.relative(repoRoot, planPath),
    matrixPath: path.relative(repoRoot, matrixPath),
    matrix,
    outPath
  };
  const startedAt = Date.now();
  const tierResults = runTiers(plan, repoRoot, { reportContext });
  const report = buildReport(reportContext.planPath, reportContext.matrixPath, matrix, tierResults);
  writeReport(outPath, report);
  const summary = report.tierSummary;
  const matrixTracking = report.matrixTracking;

  console.log('Tiered C++98 Report');
  console.log(`- Matrix weighted coverage: ${matrix.weightedPct}% (${matrix.done} done, ${matrix.partial} partial, ${matrix.missing} missing)`);
  console.log(`- Tier 1: ${summary.tier1.ok}/${summary.tier1.total} passed`);
  console.log(`- Tier 2: ${summary.tier2.ok}/${summary.tier2.total} passed`);
  console.log(`- Tier 3: ${summary.tier3.ok}/${summary.tier3.total} passed`);
  console.log(`- Matrix families tracked by tier cases: ${matrixTracking.trackedFamilies}/${matrixTracking.totalFamilies} (${matrixTracking.trackedPct}%)`);
  console.log(`- Total duration: ${formatDurationMs(Date.now() - startedAt)}`);
  console.log(`- Report JSON: ${path.relative(repoRoot, outPath)}`);

  if (summary.tier1.failed > 0 || summary.tier2.failed > 0) process.exit(1);
  process.exit(0);
}

main();
